File: upnext-ios/UpNext/UpNext/Models/BagInsights.swift
import Foundation

enum BagInsights {
    struct Change {
        let equipped: [EquipSlot: Equipment]
        let inventory: [Equipment]
        let delta: [StatKey: Int]
    }
    struct Suggestion {
        let placement: BagPlacement
        let delta: [StatKey: Int]
    }

    static func difference(_ after: [StatKey: Int], _ before: [StatKey: Int]) -> [StatKey: Int] {
        var delta: [StatKey: Int] = [:]
        for key in StatKey.allCases {
            let value = (after[key] ?? 0) - (before[key] ?? 0)
            if value != 0 { delta[key] = value }
        }
        return delta
    }

    static func loadoutStats(_ equipped: [EquipSlot: Equipment], _ inventory: [Equipment], _ rows: Int) -> [StatKey: Int] {
        var total = UpHeroBag.computeBagSynergy(equipped: equipped, inventory: inventory, rows: rows).bonuses
        for slot in UpHeroBag.anchorOrder {
            guard let item = equipped[slot] else { continue }
            for key in StatKey.allCases { total[key, default: 0] += item.stats[key] ?? 0 }
        }
        return total
    }

    static func equipmentChange(_ item: Equipment, worn: Bool, equipped: [EquipSlot: Equipment], inventory: [Equipment], rows: Int) -> Change {
        var nextEquipped = equipped
        let nextInventory: [Equipment]
        if worn {
            nextEquipped.removeValue(forKey: item.type)
            nextInventory = UpHeroBag.placeIntoBag(inventory, item, rows: rows)
        } else {
            let old = equipped[item.type]
            nextEquipped[item.type] = UpHeroBag.withoutPlacement(item)
            if let old {
                nextInventory = inventory.map { $0.id == item.id ? UpHeroBag.inheritPlacement(from: item, to: old) : $0 }
            } else {
                nextInventory = inventory.filter { $0.id != item.id }
            }
        }
        return Change(equipped: nextEquipped, inventory: nextInventory,
                      delta: difference(loadoutStats(nextEquipped, nextInventory, rows), loadoutStats(equipped, inventory, rows)))
    }

    // Deterministic improvement without reducing any stat or moving another item.
    static func suggest(_ item: Equipment, equipped: [EquipSlot: Equipment], inventory: [Equipment], rows: Int) -> Suggestion? {
        guard inventory.contains(where: { $0.id == item.id }) else { return nil }
        let layout = UpHeroBag.normalizeBagLayout(inventory, rows: rows).layout
        let before = UpHeroBag.computeBagSynergy(equipped: equipped, inventory: inventory, rows: rows).bonuses
        let rot = UpHeroBag.normalizeRot(item.bagRot)
        let rotations = UpHeroBag.canRotate(type: item.type) ? [rot, (rot + 1) % 4] : [rot]
        var best: Suggestion?
        for r in rotations {
            for y in 0..<rows {
                for x in 0..<UpHeroBag.cols {
                    guard UpHeroBag.checkPlacement(occ: layout.occupancy, rows: rows, type: item.type, x: x, y: y, rot: r, ignoreId: item.id) == .ok else { continue }
                    let p = BagPlacement(x: x, y: y, rot: r)
                    let next = inventory.map { $0.id == item.id ? UpHeroBag.withPlacement($0, p) : $0 }
                    let delta = difference(UpHeroBag.computeBagSynergy(equipped: equipped, inventory: next, rows: rows).bonuses, before)
                    guard !delta.isEmpty, delta.values.allSatisfy({ $0 > 0 }) else { continue }
                    if let current = best {
                        guard StatKey.allCases.allSatisfy({ (delta[$0] ?? 0) >= (current.delta[$0] ?? 0) }),
                              StatKey.allCases.contains(where: { (delta[$0] ?? 0) > (current.delta[$0] ?? 0) }) else { continue }
                    }
                    best = Suggestion(placement: p, delta: delta)
                }
            }
        }
        return best
    }
}

File: src/lib/bagInsights.ts
import type { Equipment, Hero, HeroBaseStats } from "@/types/uphero";
import { BAG_COLS, canRotate, checkPlacement, computeBagSynergy, inheritPlacement, normalizeBagLayout, normalizeRot, placeIntoBag, withPlacement, withoutPlacement, type BagPlacement } from "./upHeroBag";

export const BAG_STAT_KEYS = ["str", "int", "vit", "dex", "agi", "crit", "slotBonus"] as const;
export type BagStats = Partial<HeroBaseStats>;

export function statDifference(after: BagStats, before: BagStats): BagStats {
  const delta: BagStats = {};
  for (const key of BAG_STAT_KEYS) {
    const value = (after[key] ?? 0) - (before[key] ?? 0);
    if (value) delta[key] = value;
  }
  return delta;
}

/** Equipment and support only. Level growth cancels out in before/after comparisons. */
export function loadoutStats(equipped: Hero["equipped"], inventory: Equipment[], rows: number): BagStats {
  const total = { ...computeBagSynergy(equipped, inventory, rows).bonuses };
  for (const item of Object.values(equipped)) {
    if (!item) continue;
    for (const key of BAG_STAT_KEYS) total[key] = (total[key] ?? 0) + (item.stats[key] ?? 0);
  }
  return total;
}

/** Mirrors store swaps, including the old equipment inheriting the vacated footprint. */
export function equipmentChange(item: Equipment, worn: boolean, equipped: Hero["equipped"], inventory: Equipment[], rows: number) {
  const nextEquipped = { ...equipped };
  let nextInventory: Equipment[];
  if (worn) {
    delete nextEquipped[item.type];
    nextInventory = placeIntoBag(inventory, item, rows);
  } else {
    const old = equipped[item.type];
    nextEquipped[item.type] = withoutPlacement(item);
    nextInventory = old
      ? inventory.map(i => i.id === item.id ? inheritPlacement(item, old) : i)
      : inventory.filter(i => i.id !== item.id);
  }
  return { equipped: nextEquipped, inventory: nextInventory,
    delta: statDifference(loadoutStats(nextEquipped, nextInventory, rows), loadoutStats(equipped, inventory, rows)) };
}

export interface BagSuggestion { placement: BagPlacement; delta: BagStats }

/** Deterministic, non-dominated improvement with no stat regression. Never repacks other items. */
export function suggestBagPlacement(item: Equipment, equipped: Hero["equipped"], inventory: Equipment[], rows: number): BagSuggestion | null {
  if (!inventory.some(i => i.id === item.id)) return null;
  const layout = normalizeBagLayout(inventory, rows).layout;
  const before = computeBagSynergy(equipped, inventory, rows).bonuses;
  const rot = normalizeRot(item.bagRot);
  const rotations = canRotate(item.type) ? [rot, (rot + 1) % 4] : [rot];
  let best: BagSuggestion | null = null;
  for (const r of rotations) for (let y = 0; y < rows; y++) for (let x = 0; x < BAG_COLS; x++) {
    if (checkPlacement(layout.occupancy, rows, item.type, x, y, r, item.id) !== "ok") continue;
    const placement = { x, y, rot: r };
    const next = inventory.map(i => i.id === item.id ? withPlacement(i, placement) : i);
    const delta = statDifference(computeBagSynergy(equipped, next, rows).bonuses, before);
    const values = Object.values(delta);
    if (!values.length || values.some(v => v < 0)) continue;
    // Prefer a strictly better result without assigning arbitrary weights to different stats.
    if (!best || (BAG_STAT_KEYS.every(key => (delta[key] ?? 0) >= (best!.delta[key] ?? 0))
      && BAG_STAT_KEYS.some(key => (delta[key] ?? 0) > (best!.delta[key] ?? 0)))) {
      best = { placement, delta };
    }
  }
  return best;
}
